addons/certificate/models/certificate.py

```diff
@@ -0,0 +1,336 @@
+            if certificate.content_format == 'pkcs12':
+                content = certificate.with_context(bin_size=False).content
+                pkcs12_password = certificate.pkcs12_password.encode('utf-8') if certificate.pkcs12_password else None
+                key, _cert, _additional_certs = pkcs12.load_key_and_certificates(base64.b64decode(content), pkcs12_password)
+
+                if key:
+                    pem_key = base64.b64encode(key.private_bytes(
+                        encoding=Encoding.PEM,
+                        format=serialization.PrivateFormat.PKCS8,
+                        encryption_algorithm=serialization.NoEncryption()
+                    ))
+                    key_id = content_to_key_id.get((pem_key, certificate.company_id.id))
+                    if not key_id:
+                        key_id = self.env['certificate.key'].create({
+                            'name': (certificate.subject_common_name or certificate.name or "") + ".key",
+                            'content': pem_key,
+                            'company_id': certificate.company_id.id,
+                        })
+                    certificate.private_key_id = key_id
+
+    @api.depends('content', 'pkcs12_password')
+    def _compute_pem_certificate(self):
+        for certificate in self:
+            content = certificate.with_context(bin_size=False).content
+
+            if not content:
+                certificate.pem_certificate = None
+                certificate.subject_common_name = None
+                certificate.content_format = None
+                certificate.date_start = None
+                certificate.date_end = None
+                certificate.serial_number = None
+                certificate.loading_error = ""
+
+            else:
+                content = base64.b64decode(content)
+                cert = None
+
+                # Try to load the certificate in different format starting with DER then PKCS12 and
+                # finally PEM. If none succeeded, we report an error.
+                try:
+                    cert = x509.load_der_x509_certificate(content)
+                    certificate.content_format = 'der'
+                except ValueError:
+                    pass
+                if not cert:
+                    try:
+                        pkcs12_password = certificate.pkcs12_password.encode('utf-8') if certificate.pkcs12_password else None
+                        _key, cert, _additional_certs = pkcs12.load_key_and_certificates(content, pkcs12_password)
+                        certificate.content_format = 'pkcs12'
+                    except ValueError:
+                        pass
+                if not cert:
+                    try:
+                        cert = x509.load_pem_x509_certificate(content)
+                        certificate.content_format = 'pem'
+                    except ValueError:
+                        pass
```

Comment:
IMHO. This could be extracted into a function (module-level) and use `with suppress(ValueError):` + `return (cert, format)`. That avoids the nested if/try plus the except+pass.

Reply:
In any other case, yes. But here, since we manage complex security objects, the idea is to keep them inside a function and not return them. I made sure to only return odoo objects and simple python objects all throughout this PR and doing the change you suggest would defeat that purpose.

Reply:
The function could be local to the method. But in any case I don't see how a _module-level_ -- not a method-- function could cause security issues. For security sensitive functions the recommended way is to put them in the module. In any case this was a small suggestion ;)